lib/gmail-adapter.js

```diff
@@ -16,42 +17,87 @@ var Read = Juttle.proc.base.extend({
+        if (opts.length === 0) {
+            // Waiting on a juttle release that incorporates https://github.com/juttle/juttle/pull/108
+            //throw JuttleErrors.syntaxError('RT-MISSING-TIMERANGE-ERROR', {location: location});
+            throw new Error('Error: One of -from, -to, or -last must be specified to define a query time range');
+        }
+
+        // If 'from'/'to' are present, 'last' can not be present.
+        if ((_.has(options, 'from') || _.has(options, 'to')) &&
+            _.has(options, 'last')) {
+            throw JuttleErrors.syntaxError('RT-LAST-FROM-TO-ERROR', {location: location});
+        }
+
+        // 'from' must be before 'to'
+        if (_.has(options, 'from') && _.has(options, 'to') &&
+            options.from > options.to) {
+            throw JuttleErrors.syntaxError('RT-TO-FROM-MOMENT-ERROR', {location: location});
+        }
+
+        // If 'last' is specified, set appropriate 'from'/'to'
+        if (_.has(options, 'last')) {
+            this.from = JuttleMoment.now().subtract(options.last);
+            this.to = JuttleMoment.now();
+        } else {
+            // Initialize from/to if necessary.
+            this.from = options.from || JuttleMoment.now();
+            this.to = options.to || JuttleMoment.now();
         }
 
         this.raw = options.raw;
 
-        this.logger.info("Filter Expression:", this.raw);
+        this.delay = options.delay || JuttleMoment.duration(1, 's');
 
         Promise.promisifyAll(this.gmail.users.messages);
-        this.options = options;
     },
 
     start: function() {
         var self = this;
 
+        return self.get_messages_for_timerange(self.from, self.to);
+    },
+
+    teardown: function() {
+    },
+
+    // Non juttle proc methods below here
+    get_messages_for_timerange: function(from, to) {
+        this.logger.debug('get_messages_for_timerange from=' + from + " to=" + to);
+        var self = this;
+
         // Build a search expression given the options.
-        var search = self.raw;
+        var search = self.raw || "";
+
+        var now = new JuttleMoment();
 
         // The gmail search expression only has per-day granularity,
         // so we quantize the -from and -to to the nearest day and do
         // additional filtering in get_messages().
 
         // XXX/mstemm not sure how the time zone is set. It's definitely not in UTC.
+        search += " after:" + JuttleMoment.format(from, "YYYY/MM/DD", "US/Pacific");
 
-        if (self.options.from) {
-            search += " after:" + JuttleMoment.format(self.options.from, "YYYY/MM/DD", "US/Pacific");
-        }
-
-        if (self.options.to) {
-            search += " before:" + JuttleMoment.format(self.options.to, "YYYY/MM/DD", "US/Pacific");
+        // Add 1 day to to and then quantize to a day. This
+        // results in the next day, such that the before: date is
+        // always in the future.
```

Comment:
this date is a part of the "before:<date>" clause of the gmail search expression, which selects messages less than the given date. so if the actual -to time is sometime during the day of 2016-01-05, you need to select everything before:2016-01-06 so you get all messages for 2016-01-05. A second pass on the messages will compare the -to and the message receipt time to select the exact messages.